cpp/src/ts/batched_kalman.cu

```diff
@@ -0,0 +1,820 @@
+      // P = tmpB + RRT
+      for (int tid = 0; tid < N * N; tid++) {
+        l_P[tid] = l_tmpB[tid] + l_RRT[tid];
+      }
+    }
+    sum_logFs[bid] = bid_sum_logFs;
+  }
+}
+
+void batched_kalman_loop(double* ys, int nobs, const BatchedMatrix& T,
+                         const BatchedMatrix& Z, const BatchedMatrix& RRT,
+                         const BatchedMatrix& P0, const BatchedMatrix& alpha,
+                         int r, double* vs, double* Fs, double* sum_logFs) {
+  const int num_batches = T.batches();
+
+  // const int num_blocks = num_batches;
+  // const int num_threads = r * r;
+  // const size_t bytes_shared_memory = (5 * r * r + 3 * r) * sizeof(double);
+  // batched_kalman_loop_kernel<<<num_blocks, num_threads, bytes_shared_memory>>>(
+  //   ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(), r,
+  //   num_batches, vs, Fs, sum_logFs);
+  // CUDA_CHECK(cudaPeekAtLastError());
+  // CUDA_CHECK(cudaDeviceSynchronize());
+
+  dim3 numThreadsPerBlock(32, 1);
+  dim3 numBlocks(MLCommon::ceildiv<int>(num_batches, numThreadsPerBlock.x), 1);
+  if (r == 1) {
+    batched_kalman_loop_kernel_v2<1><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else if (r == 2) {
+    batched_kalman_loop_kernel_v2<2><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else if (r == 3) {
+    batched_kalman_loop_kernel_v2<3><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else if (r == 4) {
+    batched_kalman_loop_kernel_v2<4><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else if (r == 5) {
+    batched_kalman_loop_kernel_v2<5><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else if (r == 6) {
+    batched_kalman_loop_kernel_v2<6><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else if (r == 7) {
+    batched_kalman_loop_kernel_v2<7><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else if (r == 8) {
+    batched_kalman_loop_kernel_v2<8><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else {
+    throw std::runtime_error(
+      "ERROR: Currently unsupported number of parameters (r).");
+  }
+
+  CUDA_CHECK(cudaPeekAtLastError());
+  CUDA_CHECK(cudaDeviceSynchronize());
```

Comment:
device-sync is costly, please use `cudaStreamSynchronize()` instead, by passing a custom stream.